SplunkADXForwarder/app.py

```diff
@@ -0,0 +1,247 @@
+import socket
+import logging
+import time
+import os
+import yaml
+import pandas as pd
+import csv
+import threading
+import tempfile
+import shutil
+from azure.kusto.data import KustoConnectionStringBuilder
+from azure.kusto.data.data_format import DataFormat
+from azure.kusto.ingest import (
+    IngestionProperties,
+    IngestionStatus,
+    QueuedIngestClient
+)
+
+# Constants
+MAX_BATCH_SIZE = int(os.environ.get('MAX_BATCH_SIZE', '1000'))
+MAX_WAIT_TIME_SECONDS = int(os.environ.get('MAX_WAIT_TIME_SECONDS', '5'))
+MAX_RETRIES = int(os.environ.get('MAX_RETRIES', '3'))
+MAX_SIZE = 5 * 1024 * 1024 #5MB
+DELETE_TEMP_FILE_DAYS_THRESHOLD = 1
+
+def is_file_not_being_modified(file_path):
+    """
+    Returns True if the file is currently being not modified, False otherwise.
+    """
+    file_stat = os.stat(file_path)
+    current_size = file_stat.st_size
+    time.sleep(1)  # Wait for 1 second
+    new_size = os.stat(file_path).st_size
+    if current_size != new_size:
+        logging.info("the file is currently being modified")
+    return current_size == new_size
+    
+
+def ingest_data(conn_string_builder: KustoConnectionStringBuilder, database: str, table: str, table_mapping: str) -> bool:
+    """Ingest the data into Azure Data Explorer."""
+    while True:
+         # Set the folder path to read the JSON files
+        time.sleep(60)
+        folder_path = os.getcwd()
+        buffer_folder_name = os.path.join(folder_path, "ADX_Ingest")
+        logging.info(f'ingesting message from {buffer_folder_name}')
+
+        #files = os.listdir(buffer_folder_name)
+
+        ingestion_client = QueuedIngestClient(conn_string_builder)
+        ingest_props = IngestionProperties(database=database, table=table, flush_immediately=False, data_format=DataFormat.CSV, ingestion_mapping_reference=table_mapping)
+
+        if os.path.isdir(buffer_folder_name):
+            for file in get_files_not_modified(buffer_folder_name):
+                file_path = os.path.join(buffer_folder_name, file)
+                logging.info(f"Processing file :: {file_path}")
+                if file.endswith('.csv') and is_file_not_being_modified(file_path):
+                    # ingest from file
+                    df = pd.read_csv(file_path)
+                    result = ingestion_client.ingest_from_dataframe(df, ingestion_properties=ingest_props)
+                    #if ingestion result is success or queued, then move the file to temp folder
+                    if result.status == IngestionStatus.QUEUED or result.status == IngestionStatus.SUCCESS:
```

Comment:
Yes thats possible, but in Queued to Failure status changes cases the status will be visible in ingestionfailures. Actually having a getIngestionStatus in SDK would have been appropriate in this case, where we keep polling till the status is success or failure. In our case, if I only move files to tmp folder for ingestion status as SUCCESS, then the queued ingestion status files would be retried and will lead to duplicates. Hence thought of taking this approach, What do you think @ag-ramachandran?

Reply:
That should be okay, but would the following scenario occur

File is waiting for ingestion (queued status), another flush happens and the same file is sent in (this may result in duplicates), unless the scenario i understood wrongly

If we assign a SourceId , we should be able to get the status of the ingest

Reply:
I think it might occur. File identified as not being currently modified -> File queued for ingestion, ingestion status set to QUEUED -> file moves to tmp folder -> later ingestion to Kusto fails due to mapping error (lets assume).

In Kusto python SDK, I couldn't find the KustoIngestStatusTable but we do have KustoIngestStatusQueues, which may not help for high volume ingestion (https://github.com/Azure/azure-kusto-python/blob/59e263b17716b1499e596d667c1137598b98aac0/azure-kusto-ingest/azure/kusto/ingest/status.py#L50)

Reply:
may be hold a flag that says which files are under processing (in a thread safe set) and then exclude them from getting sent again and again ? Once you successfully process, remove it from that set ?